File: Core/src/main/java/jokatu/game/viewresolver/ViewResolver.java
package jokatu.game.viewresolver;

import jokatu.game.player.Player;
import org.jetbrains.annotations.NotNull;
import org.springframework.web.servlet.ModelAndView;

import java.text.MessageFormat;

/**
 * Determine which view a player should see when requesting a game.
 * @author steven
 */
public abstract class ViewResolver<P extends Player, G extends jokatu.identity.Identifiable<jokatu.game.GameID> & ophelia.event.observable.Observable<jokatu.game.event.GameEvent>> {

	protected final G game;

	protected ViewResolver(G game) {
		this.game = game;
	}

	public final ModelAndView getViewForPlayer(Player player) {
		P castPlayer = castPlayer(player);
		return getViewFor(castPlayer);
	}

	public ModelAndView getViewForObserver() {
		return getDefaultView();
	}

	@NotNull
	protected abstract ModelAndView getDefaultView();

	private P castPlayer(Player player) {
		Class<P> playerClass = handlesPlayer();
		if (!playerClass.isInstance(player)) {
			throw new IllegalArgumentException(
					MessageFormat.format("Player was not of the right type.  Expected {0}.", playerClass)
			);
		}
		return playerClass.cast(player);
	}

	@NotNull
	protected abstract Class<P> handlesPlayer();

	@NotNull
	protected abstract ModelAndView getViewFor(P player);
}

File: Core/src/main/kotlin/jokatu/game/viewresolver/ViewResolver.kt
package jokatu.game.viewresolver

import jokatu.game.Game
import jokatu.game.player.Player
import org.springframework.web.servlet.ModelAndView
import kotlin.reflect.KClass
import kotlin.reflect.full.cast

/**
 * Determine which view a player should see when requesting a game.
 * @author steven
 */
abstract class ViewResolver<P : Player, G : Game<P>> protected constructor(
		protected val playerClass: KClass<P>,
		protected val game: G
) {

	val viewForObserver: ModelAndView
		get() = defaultView

	protected abstract val defaultView: ModelAndView

	fun getViewForPlayer(player: Player): ModelAndView {
		val castPlayer = castPlayer(player)
		return getViewFor(castPlayer)
	}

	private fun castPlayer(player: Player): P {
		val playerClass = playerClass
		if (!playerClass.isInstance(player)) {
			throw IllegalArgumentException(
					"Player was not of the right type.  Expected ${playerClass.simpleName} but was ${player.javaClass.simpleName}."
			)
		}
		return playerClass.cast(player)
	}

	protected abstract fun getViewFor(player: P): ModelAndView
}
